File: agent/internal/tool/git_log.go
package tool

import (
	"context"
	"fmt"

	"github.com/xz-ephyr/raw-code/agent/pkg/api"
)

func gitLogTool() ToolDef {
	return ToolDef{
		Definition: api.ToolDefinition{
			Name:        "git_log",
			Description: "Show the commit log with author, date, and message for recent commits.",
			Category:    "git",
			Parameters: map[string]api.ParamDef{
				"path":  {Type: "string", Description: "Path to the git repository", Required: false, Default: "."},
				"count": {Type: "number", Description: "Number of recent commits to show", Required: false, Default: 10},
			},
		},
		Handler: func(ctx context.Context, e *Executor, params map[string]any) (any, error) {
			path, _ := params["path"].(string)
			if path == "" {
				path = "."
			}
			count, _ := params["count"].(float64)
			if count == 0 {
				count = 10
			}
			out, err := gitCmd(ctx, path,
				"log", fmt.Sprintf("--max-count=%d", int(count)),
				"--pretty=format:%h %ai %an: %s",
			)
			if err != nil {
				return nil, fmt.Errorf("git log failed: %w", err)
			}
			return map[string]any{"path": path, "count": int(count), "output": out}, nil
		},
	}
}

File: src/services/tools/git/gitLog.ts
import { z } from 'zod';
import type { ToolDef } from '../types';

export const gitLogTool: ToolDef = {
  name: 'git_log',
  description: 'Show commit logs. Returns recent commit history with hash, author, date, and message.',
  category: 'git',
  inputSchema: z.object({
    repoPath: z.string().describe('Absolute path to the git repository root.'),
    maxCount: z.number().optional().default(20).describe('Maximum number of commits to show.'),
    branch: z.string().optional().describe('Show log for a specific branch (defaults to current branch).'),
    path: z.string().optional().describe('Only show commits that touched this file path.'),
  }),
  execute: async ({ repoPath, maxCount, branch, path }) => {
    const backendUrl = import.meta.env.VITE_API_URL || 'http://localhost:3001';
    try {
      const res = await fetch(`${backendUrl}/tools`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ tool: 'git_log', params: { repoPath, maxCount, branch, path } }),
      });
      if (!res.ok) throw new Error(`Backend error (${res.status})`);
      return res.json();
    } catch (error: any) {
      if (error.message?.includes('fetch')) {
        throw new Error('Git tools require the Express backend to be running (http://localhost:3001) or the Tauri desktop app');
      }
      throw error;
    }
  },
};
